packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -544,91 +529,69 @@ subtask(TASK_COMPILE_SOLIDITY_GET_SOLC_BUILD)
       { run }
     ): Promise<SolcBuild> => {
       const compilersCache = await getCompilersDir();
-      const downloader = new CompilerDownloader(compilersCache);
+
+      const compilerPlatform = CompilerDownloader.getCompilerPlatform();
+      const downloader = CompilerDownloader.getConcurrencySafeDownloader(
+        compilerPlatform,
+        compilersCache
+      );
 
       const isCompilerDownloaded = await downloader.isCompilerDownloaded(
         solcVersion
       );
 
-      const { longVersion, platform: desiredPlatform } =
-        await downloader.getCompilerBuild(solcVersion);
+      if (!isCompilerDownloaded) {
+        await run(TASK_COMPILE_SOLIDITY_LOG_DOWNLOAD_COMPILER_START, {
+          solcVersion,
+          isCompilerDownloaded,
+          quiet,
+        });
 
-      await run(TASK_COMPILE_SOLIDITY_LOG_DOWNLOAD_COMPILER_START, {
-        solcVersion,
-        isCompilerDownloaded,
-        quiet,
-      });
+        await downloader.downloadCompiler(solcVersion);
 
-      let compilerPath: string | undefined;
-      let platform: CompilerPlatform | undefined;
-      let nativeBinaryFailed = false;
+        await run(TASK_COMPILE_SOLIDITY_LOG_DOWNLOAD_COMPILER_END, {
+          solcVersion,
+          isCompilerDownloaded,
+          quiet,
+        });
+      }
 
-      const compilerPathResult = await downloader.getDownloadedCompilerPath(
-        solcVersion
-      );
+      const compiler = await downloader.getCompiler(solcVersion);
 
-      if (compilerPathResult === undefined) {
-        if (desiredPlatform === CompilerPlatform.WASM) {
-          // if we were trying to download solcjs and it failed, there's nothing
-          // we can do
-          throw new HardhatError(ERRORS.SOLC.CANT_GET_COMPILER, {
-            version: solcVersion,
-          });
-        }
+      if (compiler !== undefined) {
+        return compiler;
+      }
 
-        nativeBinaryFailed = true;
-      } else {
-        compilerPath = compilerPathResult.compilerPath;
+      log(
+        "Native solc binary doesn't work, using solcjs instead. Try running npx hardhat clean --global"
+      );
 
-        // when using a native binary, check that it works correctly
-        // it it doesn't, force the downloader to use solcjs
-        if (compilerPathResult.platform !== CompilerPlatform.WASM) {
-          log("Checking native solc binary");
+      const wasmDownloader = new CompilerDownloader(
```

Comment:
Why does this use `new CompilerDownloader` instead of `CompilerDownloader.getConcurrencySafeDownloader`?

And, if using the static method is better, maybe the constructor should be private?

Reply:
Good catch. I'll make it private after completing the tests. It may be needed.